README.md
# Camomile REST API

## Deployment

The easiest way to setup a Camomile REST API is to use `docker` and `docker-compose`.  
This was tested with `docker 1.10.2` and `docker-compose 1.6.0`.

```shell
# Download docker-compose configuration file
$ wget https://raw.githubusercontent.com/camomile-project/camomile-server/master/docker-compose.yml

# Set path to MongoDB data directory
$ export CMML_DB=/where/to/store/mongodb/files

# Set path to CAMOMILE logging directory
$ export CMML_LOGS=/where/to/store/camomile/logs

# Set path to CAMOMILE media directory
$ export CMML_MEDIA=/where/media/are/stored

# Set path to CAMOMILE file metadata directory
$ export CMML_UPLOAD=/where/to/store/upload

# Set CAMOMILE port
$ export CMML_PORT=3000

# Set CAMOMILE root password
$ export CMML_PASSWORD=roO7p4s5wOrD

# Run MongoDB and CAMOMILE REST API
$ docker-compose up -d
```

### CAMOMILE client

Now, you should be able to connect to the CAMOMILE server using the Python client:

```shell
# Install Python CAMOMILE client
$ pip install camomile
$ python
```
```python
>>> from camomile import Camomile
>>> client = Camomile("http://localhost:3000")
>>> client.login("root", password="roO7p4s5wOrD")
{u'success': u'Authentication succeeded.'}
```

### Docker automated build

Thanks to Docker automated build, Docker image`camomile/api` is always in track with latest version in branch `master`.

You can however build your own Docker image using
```
$ git clone git@github.com:camomile-project/camomile-server.git
$ cd camomile-server
$ docker build -t camomile/api .
```

## Development

```shell
# Pull CAMOMILE source code from Github repository
$ git clone git@github.com:camomile-project/camomile-server.git && \
  cd camomile-server && git checkout develop

# Run MongoDB and CAMOMILE REST API (set CMML_* variables if needed)
$ docker-compose -f docker-compose.dev.yml up

# Edit CAMOMILE source code and your changes will be reflected automatically
```

## FAQ

**Warning:** this section is not up to date.

### How can I dump the database for backup?

```
$ export CMML_DUMP=/where/to/dump
$ docker run -i -t --rm \
         --link db:mongo \
         -v $CMML_DUMP:/dump \
         mongo \
         bash -c 'mongodump --host $MONGO_PORT_27017_TCP_ADDR -o /dump'
```

### How can I restore previous backup?

```
$ docker run -i -t --rm \
         --link db:mongo \
         -v $CMML_DUMP:/dump \
         mongo \
         bash -c 'mongorestore --host $MONGO_PORT_27017_TCP_ADDR /dump'
```


CHANGELOG.md
### Version 0.8 (2017-06-21)

  - feat: add event for "annotation update"
  - fix: fix default Mongo host in CLI documentation

### Version 0.7.3 (2016-07-07)

  - fix: fix a bug that crashes the server in metadata tree management

### Version 0.7.2 (2016-06-10)

  - fix: fix "watch" timeout for subsequent clients

### Version 0.7.1 (2016-06-08)

  - fix: fix "watch" timeout
  - improve: return better HTTP status code
  - feat: add Python 3 support to functional tests
  - chore: update functional tests to rely on latest Python client
  - chore: move documentation to external dedicated repository

### Version 0.7 (2016-05-24)

  - feat: {corpus | medium | layer} metadata
  - feat: {corpus | medium | layer | queue } watch

### Version 0.6.1 (2016-02-23)

  - fix: support for `docker 1.10` and `docker-compose 0.6`

### Version 0.6 (2015-09-17)

  - feat: add permissions to queue
  - feat: non-destructive queue picking
  - feat: regular users can get the list of users and groups
  - feat: regular users can get the list of their groups
  - feat: admin users can remove any account but their own
  - feat: daily log rotation
  - feat: filter layers by {fragment|data}_type
  - feat: get number of annotations in a layer
  - feat: get number of media in a corpys
  - BREAKING: getting a queue no longer returns its elements (use new picking routes instead)
  - BREAKING: only admin users can create queues
  - BREAKING: queue push returns "success" message instead of queues
  - fix: delete medium annotations upon medium deletion

### Version 0.5.6 (2015-05-13)

  - fix: delete corpus after its layers and media are removed

### Version 0.5.4 (2015-05-12)

  - fix: fix a bug when checking for non-existing resource

### Version 0.5.3 (2015-05-07)

  - fix: fix a bug in .../annotation/... routes

### Version 0.5.2 (2015-05-05)

  - fix: increase POST limit to 50mb

### Version 0.5.1 (2015-04-30)

  - fix: only corpus administrators can create media
  - fix: only corpus readers can get list of media
  - doc: update Docker instructions

### Version 0.5 (2015-04-29)

  - major code refactoring
  - switch to ExpressJS 4
  - BREAKING CHANGE in external API
  - BREAKING CHANGE in internal database
  - wip: unit test
  - wip: documentation for Javascript and Python clients

### Version 0.4.1 (2015-02-03)

  - various code improvement
  - remove route /corpus/:corpus/medias
  - remove route /layer/:layer/annotations
  - update route POST /layer/:layer/annotation
  - defaults to no history (use ?history=ON to return history)
  - add filtering support for all GET routes

### Version 0.3.1 (2014-12-15)

  - bug fix release

### Version 0.2 (2014-12-08)

  - database overhaul
  - complete route refactoring

### Version 0.0.1 (2014-05-13)

  - first public release


app.js
/*
The MIT License (MIT)

Copyright (c) 2013-2014 CNRS

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

var http = require('http');
var express = require('express');

var morgan = require('morgan');
var fileStreamRotator = require('file-stream-rotator');
var fs = require('fs');

var cors = require('cors');
var bodyParser = require('body-parser');
var methodOverride = require('method-override');
var session = require('express-session');
var cookieParser = require('cookie-parser');

var program = require('commander');
var mongoose = require('mongoose');
var mongoStore = require('connect-mongo')(session);

var routes = require('./routes');
var User = require('./models/User');
var Authentication = require('./controllers/Authentication');

program
  .option('--port <port>', 'Local port to listen to (default: 3000)', parseInt)
  .option('--mongodb-host <host>', 'MongoDB host (default: mongo)')
  .option('--mongodb-port <port>', 'MongoDB port (default: 27017)', parseInt)
  .option('--mongodb-name <dbname>',
    'MongoDB database name (default: camomile)')
  .option('--root-password <dbname>', 'Change/set root password')
  .option('--media <dir>', 'Path to media root directory')
  .option('--log <dir>', 'Path to log directory (default: ' + __dirname + '/log)')
  .option('--upload <dir>', 'Path to upload root directory (default: ' + __dirname + '/upload)')
  .parse(process.argv);

var port =
  program.port ||
  process.env.PORT ||
  3000;

// CLI || env || linked mongo Docker container
var mongodb_host =
  program.mongodbHost ||
  process.env.MONGODB_HOST ||
  'mongo';

// CLI || env || default
var mongodb_port =
  program.mongodbPort ||
  process.env.MONGODB_PORT ||
  27017;

// CLI || env || default
var mongodb_name =
  program.mongodbName ||
  process.env.MONGODB_NAME ||
  'camomile';

var root_password =
  program.rootPassword ||
  process.env.ROOT_PASSWORD;

var media =
  program.media ||
  process.env.MEDIA ||
  '/media';

var upload =
    program.upload ||
    process.env.UPLOAD ||
    __dirname + '/upload';

// CLI || env || ./log
var logDirectory =
  program.log ||
  process.env.LOG ||
  __dirname + '/log';

// env || random
var cookieSecret =
  process.env.COOKIE_SECRET ||
  Authentication.helper.cookieSecret();

var app = express();

app.set('port', port);
app.set('media', media);
app.set('upload', upload);

// === LOGGING ================================================================

// create log directory
fs.existsSync(logDirectory) || fs.mkdirSync(logDirectory);

// create a (daily) rotating write stream
var accessLogStream = fileStreamRotator.getStream({
  filename: logDirectory + '/%DATE%.log',
  frequency: 'daily',
  verbose: false,
  date_format: 'YYYYMMDD'
});

morgan.token('user', function (req, res) {
  if (req.session && req.session.user) {
    return req.session.user.username;
  } else {
    return 'anonymous';
  };
});

var logFormat = '[:date[clf]] :user (:remote-addr) :method :url :status :response-time ms';

var logger = morgan(logFormat, {
  skip: function (req, res) {
    return req.method === 'GET' || req.method === 'OPTIONS';
  },
  stream: accessLogStream
});

app.use(logger);

// === CROSS-ORIGIN RESOURCE SHARING ==========================================

var cors_options = {
  origin: true,
  methods: ['GET', 'PUT', 'POST', 'DELETE'],
  allowedHeaders: ['X-CSRF-Token', 'X-Requested-With', 'Accept',
    'Accept-Version', 'Content-Length', 'Content-MD5', 'Content-Type', 'Date',
    'X-Api-Version'
  ],
  credentials: true,
};

app.use(cors(cors_options));

// ============================================================================

app.use(methodOverride());

// === SESSION ================================================================

mongoose.connect('mongodb://' + mongodb_host + ':' + mongodb_port + '/' +
  mongodb_name);
mongoose.set('debug', false);

var sessionStore = new mongoStore({
  mongooseConnection: mongoose.connection,
  db: mongoose.connections[0].db,
  clear_interval: 60
});

app.use(session({
  cookie: {
    secure: false,
    maxAge: 24 * 60 * 60 * 1000,
  },
  name: 'camomile.sid',
  proxy: true,
  resave: false,
  rolling: true,
  saveUninitialized: false,
  secret: cookieSecret,
  store: sessionStore,
}));

app.use(cookieParser(cookieSecret));

// ============================================================================

app.use(bodyParser.json({
  limit: '50mb'
}));
app.use(bodyParser.urlencoded({
  limit: '50mb',
  extended: true
}));

// ============================================================================

// handle CORS pre-flight requests
// (must be added before any other route)
app.options('*', cors(cors_options));

//start routes:
routes.initialize(app);

User.findOne({
  username: "root"
}, function (error, root) {

  if (!root) {

    if (!root_password) {
      console.log('Please set root password. Exiting.');
      process.exit(-1);
    }

    root = new User({
      username: 'root',
      role: 'admin'
    });

  }

  if (root_password) {
    Authentication.helper.generateSaltAndHash(root_password, function (error, salt,
      hash) {
      root.salt = salt;
      root.hash = hash;
      root.save(function (error) {
        if (error) {
          console.log('Could not set root password. Exiting.');
          process.exit(-1);
        }
        console.log('Root password successfully set.');
      });
    });
  }
});

app.listen(app.get('port'), function () {
  console.log('Server listening on port ' + app.get('port'));
});
